torch/_functorch/_aot_autograd/traced_function_transforms.py

```diff
@@ -235,22 +236,45 @@ def inner_fn(primals: List[Any], tangents: List[Any]):
         backward_out: Tuple[Tensor, ...] = ()
         # Call the backwards pass
         if grad_primals:
-            with fx_traceback.preserve_node_meta():
-                # for full graph export, we always export a joint graph where we assume no tangents are needed.
-                if aot_config.no_tangents:
-                    assert len(needed_tangents) == 1 and needed_tangents[0].numel() == 1
-                    backward_out = torch.autograd.grad(
-                        needed_outs,
-                        grad_primals,
-                        allow_unused=True,
+            with contextlib.ExitStack() as context_stack:
+                functional_tensor_mode = (
+                    torch.utils._python_dispatch._detect_infra_mode(
+                        torch._C._TorchDispatchModeKey.FUNCTIONAL
                     )
-                else:
-                    backward_out = torch.autograd.grad(
-                        needed_outs,
-                        grad_primals,
-                        grad_outputs=needed_tangents,
-                        allow_unused=True,
+                )
+                if functional_tensor_mode is not None:
+                    # Prevent partitioner from moving effectful ops happened in backward to forward for side-effects correctness.
```

Comment:
nit: this comment applies to your use of `set_partitioner_tag` further down, not to the setting of `_tokens_forward_outputs`